### ISSUE TITLE
Is it possible to apply algo procedures on sub-graphs ?

### ISSUE THREAD

**Issue by ajkl** [neo4j-contrib/neo4j-graph-algorithms#326]
Is there a way to get a subgraph using cypher and apply the graph algorithms say betweenness-centrality to the subgraph only ?

**Comment by jexp** [comment]
Yes. You can use the `graph:'cypher'` variant for that. There are examples in the readme and docs.
Let us know if that works for you.

**Comment by ajkl** [comment]
Duh! My bad, should have read the docs better. Got it. Worked. 
Next question - If I am not wrong the centrality algorithms work only on fixed node types ? Can I run them on mixed set of node types returned by a cypher query.
eg: I want to run a betweenness centrality on all nodes of this cypher MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains 'The Matrix' return *
Using this dataset to play - https://github.com/neo4j-examples/cineasts-spring-data-neo4j

PS: do let me know if there is a better forum you guys prefer for such newbie questions.

**Comment by jexp** [comment]
Cool, glad it worked out. If we can improve the docs please let us know.

Btw. the algorithms internally don't care about node or relationship-types.

If all is ok, please  close the issue.

**Comment by ajkl** [comment]
I am trying to run this but I get centrality score 0 on all. What am I missing
`CALL algo.betweenness.stream(
'MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains "The Matrix" RETURN id(n) AS id UNION MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains "The Matrix" RETURN id(a) AS id',
'MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) RETURN id(a) as source, id(n) as target',
{graph:'cypher', write: true})
YIELD nodeId, centrality
RETURN nodeId,centrality order by centrality desc`

**Comment by tomasonjo** [comment]
What are you trying to analyse ?

The network that you are analysing right now looks like this (from the `:play movies` sample)



![movies](https://user-images.githubusercontent.com/19948365/29740327-2b2021da-8a54-11e7-863a-2b028cf4ace4.png)

Remember that the betweenness algorithm is designed for directed networks for now. You have relationships pointing from actor to movie, and and nothing further. So there is no node on the path between different nodes, hence the betweenness centrality is always 0. Depends what you want to achieve, but here is an example for the `:play movies` dataset, that analyses a network between actors based on the same movies they played in. If two actors played in the same movie, a virtual relationship is created between them. 

```
CALL algo.betweenness.stream(
'MATCH (p:Person) RETURN id(p) as id',
'MATCH (a1:Person)-[:ACTED_IN]->(n:Movie)<-[:ACTED_IN]-(a:Person) RETURN id(a) as source, id(a1) as target',
{graph:'cypher'})
YIELD nodeId, centrality
RETURN nodeId,centrality order by centrality desc
```

**Comment by ajkl** [comment]
Aah I missed the part that it's only for directed graphs! Are there plans
to support a way to ignore the direction?

**Comment by jexp** [comment]
@ajkl can you put in a separate issue for undirected graphs?
In general with the cypher loading you can specify your load query in a way that loads relationships of both directions between nodes.

**Comment by tomasonjo** [comment]
There are now 2 issues for undirected graph one is [mine](https://github.com/neo4j-contrib/neo4j-graph-algorithms/issues/148) and one is from [@ajkl](https://github.com/neo4j-contrib/neo4j-graph-algorithms/issues/331).

Based on michaels advice you can change you query to this to make it "undirected".

```
CALL algo.betweenness.stream(
'MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains "The Matrix" RETURN id(n) AS id UNION MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains "The Matrix" RETURN id(a) AS id',
'MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) RETURN id(a) as source, id(n) as target UNION MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) RETURN id(n) as source, id(a) as target ',
{graph:'cypher', write: true})
YIELD nodeId, centrality
RETURN nodeId,centrality order by centrality desc
```
@ajkl please let us know how this works out.

**Comment by ajkl** [comment]
@jexp I had closed this one and created https://github.com/neo4j-contrib/neo4j-graph-algorithms/issues/331
. Also https://github.com/neo4j-contrib/neo4j-graph-algorithms/issues/148 is similar to the one i created, @tomasonjo we can close mine if you like.
I see this solution would work, but the UNION way seems a bit hacky to me. I still feel something like a flag would make things simpler.

**Comment by tomasonjo** [comment]
@ajkl 
Yes I agree with you. And we need it also when we are loading labels and relationships from graph, this "hack" works only for cypher loading.

**Comment by ajkl** [comment]
@tomasonjo @jexp Although the cypher trick works for making it direction agnostic, i think we face the issue mentioned in here https://github.com/neo4j-contrib/neo4j-graph-algorithms/issues/282
If I try this 
```
CALL algo.betweenness.stream(
'MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains "The Matrix" RETURN id(n) AS id UNION MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) where n.title contains "The Matrix" RETURN id(a) AS id',
'MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) RETURN id(a) as source, id(n) as target UNION MATCH (n:Movie)<-[:ACTS_IN]-(a:Actor) RETURN id(n) as source, id(a) as target ',
{graph:'cypher', write: true})
YIELD nodeId, centrality
RETURN nodeId,centrality order by centrality desc
```
only `Keanu Reeves` gets all the love :) Thats the only person/actor who gets a non-zero centrality.
```
╒════════╤══════════════════╕
│"nodeId"│"centrality"      │
╞════════╪══════════════════╡
│461     │4944.613636363636 │
├────────┼──────────────────┤
│448     │4850.545454545454 │
├────────┼──────────────────┤
│431     │3552.5499999999997│
├────────┼──────────────────┤
│434     │630.8136363636362 │
├────────┼──────────────────┤
│59386   │1.5               │
├────────┼──────────────────┤
│447     │0                 │
├────────┼──────────────────┤
│446     │0                 │
├────────┼──────────────────┤
│445     │0                 │
├────────┼──────────────────┤
│444     │0                 │
├────────┼──────────────────┤
│443     │0                 │
├────────┼──────────────────┤
│442     │0                 │
├────────┼──────────────────┤
│441     │0                 │
├────────┼──────────────────┤
│440     │0                 │
├────────┼──────────────────┤
│439     │0                 │
├────────┼──────────────────┤
│438     │0                 │
├────────┼──────────────────┤
│437     │0                 │
├────────┼──────────────────┤
│436     │0                 │
├────────┼──────────────────┤
│435     │0                 │
├────────┼──────────────────┤
│460     │0                 │
├────────┼──────────────────┤
│459     │0                 │
├────────┼──────────────────┤
│458     │0                 │
├────────┼──────────────────┤
│457     │0                 │
├────────┼──────────────────┤
│456     │0                 │
├────────┼──────────────────┤
│455     │0                 │
├────────┼──────────────────┤
│454     │0                 │
├────────┼──────────────────┤
│453     │0                 │
├────────┼──────────────────┤
│452     │0                 │
├────────┼──────────────────┤
│451     │0                 │
├────────┼──────────────────┤
│450     │0                 │
├────────┼──────────────────┤
│449     │0                 │
├────────┼──────────────────┤
│465     │0                 │
├────────┼──────────────────┤
│464     │0                 │
├────────┼──────────────────┤
│463     │0                 │
├────────┼──────────────────┤
│462     │0                 │
└────────┴──────────────────┘
```

**Comment by tomasonjo** [comment]
If you are using the latest release, then this is the case. You can build yourself the latest version and it should work fine as this bug was fixed in the meantime.

**Comment by mknblch** [comment]
I just refactored the betweenness procedures. It now accepts the direction used for its BFS as parameter in the config-map. Its implemented in PR #345. Once it is merged we should test if the values makes sense.
bets regards

**Comment by mneedham** [comment]
Closing as #345 is now merged

**Comment by tiffanyol** [comment]
hi @ajkl 
When you said,
> First 4 are the movie nodes and 59386 is Keanu Reeves

If i want to find highest centrality movies only, what is the best way to distinguish between Actor and Movie nodes in the betweenness result? thanks!
